dask/array.py

```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
+        return {index/blocksize : index % blocksize}
+
+    elif isinstance(index, slice):
+        #How do we deal with a step being bigger than the block?
+        #How do we deal with a negative slice?
+        start = index.start or 0
+        stop = index.stop or dim_shape
+        step = index.step
+        start_block = start/blocksize
+        stop_block = int(ceil(stop/blocksize))+1
+
+        res = [_block_slice_step_start(blocksize, block, start, stop, step) for block in range(0, stop_block)]
+                    
+        return {block: res[block] for block in range(start_block, stop_block) if res[block] is not None}
+
+                       
+
+
+def _block_slice_step_start(blocksize, blocknum, start, stop, step):
+    """
+    Parameters
+    ----
+    blocksize : integer
+      Denotes the size of the current block
+    blocknum : integer
+      Denotes the index of the block (block indexing starts from 0)
+      Example:
+        - blocksize = 25
+        - blocknum = 0
+        This implies that this block (0) holds block indexes [0:25] and absolute indexes [0:25]
+
+        - blocksize = 25
+        - blocknum = 3
+        This implies that this block (3) holds block indexes [0:25] and absolute indexes [75:100]
+        
+    start : integer
+      Denotes the absolute starting index (inclusive) of the slice
+      start is NOT the start within this block
+      start IS the start index for the entire (aka non-blocked) slice
+    stop : integer
+      Denotes the absolute stopping index (exclusive) of the slice
+      stop is NOT the stop index within this block
+      stop IS the stop index for the entire slice
+    step : An integer or None
+      Denotes the step size of the slice
+
+
+    Returns
+    ----
+    - None if start falls after the highest absolute index in the block
+    - None if stop falls before the lowest absolute index in the block
+    - None if the step size is so large that the current block won't contribute any elements
+    - An integer index if the step size is large enough that the current block only contributes a single element (in some cases)
+    - a slice() containing the correct start and stop indexes into the block plus the step if any
+    
+    Notes
+    ----        
+    - if step > blocksize, we are doing basic indexing into the blocks and may end up skipping a block
+    - negative stepping is UNHANDLED as of 2015-01-16
+
+    >>> _block_slice_step_start(20, 0, 10, 20, None)
+    slice(10, 20, None)
+    >>> _block_slice_step_start(20, 0, 20, 50, None) #The starting index falls outside the block, so it returns None
+    >>> _block_slice_step_start(20, 1, 10, 50, None) #The starting index falls in block 0, so this block's slice starts at index 0
+    slice(0, 20, None)
+    >>> _block_slice_step_start(20, 1, 10, 20, None) #The stopping index falls before block 1 so it returns None
+    >>> _block_slice_step_start(20, 0, 3, 100, 25)
```

Comment:
Hrm, it looks like in the case of uneven step sizes some blocks might be indexed with slices while some might be indexed with integers.  These don't return the same kind of thing when passed to numpy arrays.

```Python
In [4]: x[3:20:25]
Out[4]: array([ 1.])

In [5]: x[3]
Out[5]: 1.0
```

So it might be wise to return a slice regardless.